File: clients/java/src/test/java/io/volant/ScramAdminTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** Create/Delete/ListScramUsers client tests against a fake native server. */
class ScramAdminTest {
    @Test
    void createOk() throws Exception {
        try (ScramAdminServer srv = new ScramAdminServer(0, 0, 0, Collections.emptyList())) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.createScramUser("alice", "s3cret", 4096);
            }
            assertEquals("alice", srv.createUser.get());
            assertEquals("s3cret", srv.createPass.get());
            assertEquals(4096, srv.createIters.get());
            assertEquals(Collections.singletonList(Codec.OP_CREATE_SCRAM_USER), srv.opcodes);
        }
    }

    @Test
    void deleteNotFoundRaises() throws Exception {
        try (ScramAdminServer srv = new ScramAdminServer(0, 2, 0, Collections.emptyList())) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.deleteScramUser("missing"));
                assertEquals(2, ex.code);
                assertEquals("delete_scram_user", ex.op);
            }
            assertEquals("missing", srv.deleteUser.get());
        }
    }

    @Test
    void listReturnsNames() throws Exception {
        try (ScramAdminServer srv = new ScramAdminServer(0, 0, 0, Arrays.asList("alice", "bob"))) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                List<String> names = c.listScramUsers();
                assertEquals(Arrays.asList("alice", "bob"), names);
            }
            assertTrue(srv.listPayloadEmpty);
        }
    }

    @Test
    void unauthorizedRaises() throws Exception {
        try (ScramAdminServer srv = new ScramAdminServer(0, 0, 23, Collections.emptyList())) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, c::listScramUsers);
                assertEquals(23, ex.code);
                assertEquals("list_scram_users", ex.op);
            }
        }
    }

    private static final class ScramAdminServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> createUser = new AtomicReference<>();
        final AtomicReference<String> createPass = new AtomicReference<>();
        final AtomicInteger createIters = new AtomicInteger();
        final AtomicReference<String> deleteUser = new AtomicReference<>();
        final List<Integer> opcodes = new CopyOnWriteArrayList<>();
        volatile boolean listPayloadEmpty;
        private final int createError;
        private final int deleteError;
        private final int listError;
        private final List<String> usernames;
        private final ServerSocket listen;
        private final Thread thread;

        ScramAdminServer(int createError, int deleteError, int listError, List<String> usernames)
                throws IOException {
            this.createError = createError;
            this.deleteError = deleteError;
            this.listError = listError;
            this.usernames = usernames;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-scram-admin");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    opcodes.add(d.frame.opcode);
                    byte[] payload;
                    int respOp;
                    if (d.frame.opcode == Codec.OP_CREATE_SCRAM_USER) {
                        Codec.CreateScramUserRequest req = Codec.decodeCreateScramUserRequest(d.frame.payload);
                        createUser.set(req.username);
                        createPass.set(req.password);
                        createIters.set(req.iterations);
                        payload = Codec.encodeCreateScramUserResponse(
                                new Codec.CreateScramUserResponse(createError));
                        respOp = Codec.OP_CREATE_SCRAM_USER_RESPONSE;
                    } else if (d.frame.opcode == Codec.OP_DELETE_SCRAM_USER) {
                        Codec.DeleteScramUserRequest req = Codec.decodeDeleteScramUserRequest(d.frame.payload);
                        deleteUser.set(req.username);
                        payload = Codec.encodeDeleteScramUserResponse(
                                new Codec.DeleteScramUserResponse(deleteError));
                        respOp = Codec.OP_DELETE_SCRAM_USER_RESPONSE;
                    } else if (d.frame.opcode == Codec.OP_LIST_SCRAM_USERS) {
                        listPayloadEmpty = d.frame.payload == null || d.frame.payload.length == 0;
                        payload = Codec.encodeListScramUsersResponse(
                                new Codec.ListScramUsersResponse(listError, usernames));
                        respOp = Codec.OP_LIST_SCRAM_USERS_RESPONSE;
                    } else {
                        return;
                    }
                    out.write(Frame.encode(respOp, d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception ignored) {
                // test assertions cover the client side
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/go/scram_admin_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type scramAdminResult struct {
	opcodes     []uint16
	createUser  string
	createPass  string
	createIters uint32
	deleteUser  string
	listPayload []byte
	err         error
}

func serveScramAdmin(t *testing.T, createErr, deleteErr, listErr uint16, names []string) (addr string, got *scramAdminResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &scramAdminResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			res.opcodes = append(res.opcodes, f.Opcode)
			var (
				payload []byte
				op      uint16
			)
			switch f.Opcode {
			case codec.OpCreateScramUser:
				req, err := codec.DecodeCreateScramUserRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.createUser = req.Username
				res.createPass = req.Password
				res.createIters = req.Iterations
				payload, err = codec.EncodeCreateScramUserResponse(codec.CreateScramUserResponse{ErrorCode: createErr})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpCreateScramUserResponse
			case codec.OpDeleteScramUser:
				req, err := codec.DecodeDeleteScramUserRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.deleteUser = req.Username
				payload, err = codec.EncodeDeleteScramUserResponse(codec.DeleteScramUserResponse{ErrorCode: deleteErr})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpDeleteScramUserResponse
			case codec.OpListScramUsers:
				res.listPayload = append([]byte(nil), f.Payload...)
				payload, err = codec.EncodeListScramUsersResponse(codec.ListScramUsersResponse{
					ErrorCode: listErr,
					Usernames: names,
				})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpListScramUsersResponse
			default:
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			raw, err := frame.Encode(op, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestCreateScramUserOk(t *testing.T) {
	addr, got, stop := serveScramAdmin(t, 0, 0, 0, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if err := c.CreateScramUser("alice", "s3cret", 4096); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.createUser != "alice" || got.createPass != "s3cret" || got.createIters != 4096 {
		t.Fatalf("create %#v", got)
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpCreateScramUser {
		t.Fatalf("opcodes %#v", got.opcodes)
	}
}

func TestCreateScramUserDefaultEncodesZeroIterations(t *testing.T) {
	addr, got, stop := serveScramAdmin(t, 0, 0, 0, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if err := c.CreateScramUserDefault("alice", "s3cret"); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.createUser != "alice" || got.createPass != "s3cret" || got.createIters != 0 {
		t.Fatalf("create %#v", got)
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpCreateScramUser {
		t.Fatalf("opcodes %#v", got.opcodes)
	}
}

func TestDeleteScramUserNotFoundRaises(t *testing.T) {
	addr, got, stop := serveScramAdmin(t, 0, 2, 0, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	err = c.DeleteScramUser("missing")
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "delete_scram_user" {
		t.Fatalf("broker error %+v", be)
	}
	if got.deleteUser != "missing" {
		t.Fatalf("delete %q", got.deleteUser)
	}
}

func TestListScramUsersReturnsNames(t *testing.T) {
	addr, got, stop := serveScramAdmin(t, 0, 0, 0, []string{"alice", "bob"})
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	names, err := c.ListScramUsers()
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if len(names) != 2 || names[0] != "alice" || names[1] != "bob" {
		t.Fatalf("names %#v", names)
	}
	if len(got.listPayload) != 0 {
		t.Fatalf("list request %x", got.listPayload)
	}
}

func TestListScramUsersUnauthorizedRaises(t *testing.T) {
	addr, _, stop := serveScramAdmin(t, 0, 0, 23, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.ListScramUsers()
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 23 || be.Op != "list_scram_users" {
		t.Fatalf("broker error %+v", be)
	}
}
